### PR TITLE
pmi: factor out handling of wire protocols

### PR DESCRIPTION
**Pull request by hzhou**
## Pull Request Description
Add utilities in `pmi_wire.c` to handle wire protocols.
Refactor `pmi_v1.c` and `pmi_v2.c` with the common wire utilities.

## NOTES 
* The `PMIU_cmd` utility interface are defined in [pmi_wire.h](https://github.com/pmodels/mpich/blob/33465bb7b899c7d8bbf9d64fdf9c86a24fe50eaa/src/pmi/src/pmi_wire.h)

* Currently it uses a fixed token array:
```
#define MAX_PMI_ARGS 1000
struct PMIU_cmd {
    char *buf;
    int len;
    int version;                /* wire protocol: 1 or 2 */
    const char *cmd;
    struct PMII_token tokens[MAX_PMI_ARGS];
    int num_tokens;
};

```
Most of the PMI command does not need that many tokens -- likely 20 or so is sufficient. But PMI_Spawn potentially can use many tokens and 1000 may not cover some odd usages. So
* TODO: add a dynamic extension to token arrays.

## Author Checklist
* [x] **Provide Description** 
      Particularly focus on _why_, not _what_. Reference background, issues, test failures, xfail entries, etc.
* [x] **Commits Follow Good Practice**
      Commits are self-contained and do not do two things at once. 
      Commit message is of the form: `module: short description` 
      Commit message explains what's in the commit.
* [x] **Passes All Tests**
      Whitespace checker. Warnings test. Additional tests via comments.
* [x] **Contribution Agreement**
      For non-Argonne authors, check [contribution agreement](http://www.mpich.org/documentation/contributor-docs/). 
      If necessary, request an explicit comment from your companies PR approval manager.

### CODE REVIEW

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

**Comment by hzhou** [comment]
test:mpich/pmi

File: src/pmi/src/pmi_wire.c [hunk 1]
```diff
@@ -240,6 +240,26 @@ void PMII_pmi_init(struct PMII_pmi *pmi, int version, const char *cmd)
     pmi->num_tokens = 0;
 }
 
+void PMII_pmi_free_buf(struct PMII_pmi *pmi)
+{
+    MPL_free(pmi->buf);
+    pmi->buf = NULL;
+}
```
Comment (raffenet) [inline src/pmi/src/pmi_wire.c:247]: The commit message says allocating the buffer supports reentrence, but are we really doing reentrent interfaces here or just multi-threaded? Using heap buffers helps us keep them thread private.

**Review by raffenet — COMMENTED** [review]
Just minor comments, I think.

File: src/pmi/src/pmi_wire.c [hunk 2]
```diff
@@ -492,22 +540,40 @@ int PMII_pmi_read(int fd, struct PMII_pmi *pmi)
...
+        const char *thrid;
+        thrid = PMII_pmi_find_keyval(pmi, "thrid");
+        if (thrid) {
+            struct PMII_pmi *cmd_pmi;
+            int rc;
+            rc = sscanf(thrid, "%p", &cmd_pmi);
+            PMIU_ERR_CHKANDJUMP1(rc != 1, pmi_errno, PMIU_FAIL, "bad thrid (%s)\n", thrid);
+
+            if (cmd_pmi != pmi) {
+                /* transfer and reset pmi->buf to NULL */
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pmi/src/pmi_wire.c:570]: We received a PMI message and has parsed it into `pmi` object, but it belongs to an existing `cmd_pmi` object. So we need transfer (or copy) the data (what we have read/received) to `cmd_pmi`.

File: src/pmi/src/pmi_wire.c (hunk 1 above)
Comment (hzhou) [reply src/pmi/src/pmi_wire.c:247]: Here by reentrant I mean the read routine can read another PMI message while a previous PMI message is still being handled. The previous PMI1 utility uses global static buffer. By allocating buffer with each PMI object, we allow multiple PMI buffers to be alive.

File: src/pmi/src/pmi_wire.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (raffenet) [reply src/pmi/src/pmi_wire.c:570]: Got it. I think this matches my understanding so nothing to do.

File: src/pmi/src/pmi_wire.h [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+};
+
+struct PMII_pmi {
+    char *buf;
+    int len;
+    int version;                /* wire protocol: 1 or 2 */
+    const char *cmd;
+    struct PMII_token tokens[MAX_PMI_ARGS];
+    int num_tokens;
+};
```
[the remark this answers is not in the document]
Comment (hzhou) [reply src/pmi/src/pmi_wire.h:35]: We discussed and we are going to name it `PMIU_cmd`.

[the remark this answers is not in the document]
Comment (hzhou) [reply src/pmi/src/pmi_wire.h]: Done namechange
```
s/PMII_pmi/PMIU_cmd/g
s/\<pmi\>/pmicmd/g
```

File: src/pmi/src/pmi_wire.c (hunk 1 above)
Comment (hzhou) [reply src/pmi/src/pmi_wire.c]: Commit message revised:

```
    We need attach the buffer within each pmi object to allow multiple
    PMIU_cmd objects concurrently. Add PMIU_cmd_free_buf to free the buffer
    once done.
```

**Review by raffenet — COMMENTED** [review]
I don't have any new comments for the changes since last review, but would like to see clean tests before digging in again.

**Comment by hzhou** [comment]
test:mpich/pmi
test:mpich/ch4/most
test:mpich/ch3/most

**Comment by hzhou** [comment]
test:mpich/ch3/tcp

File: src/pmi/src/pmi_wire.c [hunk 4]
```diff
@@ -538,3 +538,30 @@ int PMIU_cmd_send(int fd, struct PMIU_cmd *pmicmd)
...
     goto fn_exit;
 }
+
+/*
+ * This function is used to request information from the server and check
+ * that the response uses the expected command name.  On a successful
+ * return from this routine, the response is parsed into the pmi object.
+ * It can be queried for attributes.
+ */
+int PMII_pmi_get_response(int fd, struct PMII_pmi *pmi, const char *expectedCmd)
```
Comment (raffenet) [inline src/pmi/src/pmi_wire.c]: nitpick: this commit uses the old name then gets updated in the next commit

**Review by raffenet — APPROVED** [review]
One minor comment, but overall should be OK to merge.

File: src/pmi/src/pmi_wire.c (hunk 4 above)
Comment (hzhou) [reply src/pmi/src/pmi_wire.c]: Thanks, I'll fix it before merging

Comment (hzhou) [reply src/pmi/src/pmi_wire.c]: I'll simply squash the two commits together. Both are updating utility functions before actual usage.

**Pull request merged**